plugin/plugins/kafka/service/service_test.go

```diff
@@ -0,0 +1,560 @@
+	var length int
+	var err error
+	waitChan := make(chan struct{}, 0)
+	go func() {
+		length, err = iw.Write(testBytes)
+		waitChan <- struct{}{}
+	}()
+	receivedBytes := <-outChan
+	<-waitChan
+	require.Equal(t, len(testBytes), length)
+	require.Equal(t, testBytes, receivedBytes)
+	require.Nil(t, err)
+}
+
+// change this to write to in-memory io.Writer (e.g. bytes.Buffer)
+func TestKafkaStreamingService(t *testing.T) {
+	 testingCtx = sdk.NewContext(nil, types1.Header{}, false, log.TestingLogger())
+	testKeys := []types.StoreKey{mockStoreKey1, mockStoreKey2}
+	kss, err := NewKafkaStreamingService(producerConfig, topicPrefix, flushTimeoutMs, testKeys, testMarshaller, true)
+	testStreamingService = kss
+	require.Nil(t, err)
+	require.IsType(t, &KafkaStreamingService{}, testStreamingService)
+	require.Equal(t, topicPrefix, testStreamingService.topicPrefix)
+	require.Equal(t, testMarshaller, testStreamingService.codec)
+	deleteTopics(t, topics, bootstrapServers)
+	createTopics(t, topics, bootstrapServers)
+	testListener1 = testStreamingService.listeners[mockStoreKey1][0]
+	testListener2 = testStreamingService.listeners[mockStoreKey2][0]
+	wg := new(sync.WaitGroup)
+	testStreamingService.Stream(wg)
+	testListenBeginBlock(t)
+	testListenDeliverTx1(t)
+	testListenDeliverTx2(t)
+	testListenEndBlock(t)
+	testStreamingService.Close()
+	wg.Wait()
+}
+
+func  testListenBeginBlock(t *testing.T) {
+	expectedBeginBlockReqBytes, err := testMarshaller.MarshalJSON(&testBeginBlockReq)
+	require.Nil(t, err)
+	expectedBeginBlockResBytes, err := testMarshaller.MarshalJSON(&testBeginBlockRes)
+	require.Nil(t, err)
+
+	// write state changes
+	testListener1.OnWrite(mockStoreKey1, mockKey1, mockValue1, false)
+	testListener2.OnWrite(mockStoreKey2, mockKey2, mockValue2, false)
+	testListener1.OnWrite(mockStoreKey1, mockKey3, mockValue3, false)
+
+	// expected KV pairs
+	expectedKVPair1, err := testMarshaller.MarshalJSON(&types.StoreKVPair{
+		StoreKey: mockStoreKey1.Name(),
+		Key:      mockKey1,
+		Value:    mockValue1,
+		Delete:   false,
+	})
+	require.Nil(t, err)
+	expectedKVPair2, err := testMarshaller.MarshalJSON(&types.StoreKVPair{
+		StoreKey: mockStoreKey2.Name(),
+		Key:      mockKey2,
+		Value:    mockValue2,
+		Delete:   false,
+	})
+	require.Nil(t, err)
+	expectedKVPair3, err := testMarshaller.MarshalJSON(&types.StoreKVPair{
+		StoreKey: mockStoreKey1.Name(),
+		Key:      mockKey3,
+		Value:    mockValue3,
+		Delete:   false,
+	})
+	require.Nil(t, err)
+
+	// send the ABCI messages
+	err = testStreamingService.ListenBeginBlock(testingCtx, testBeginBlockReq, testBeginBlockRes)
+	require.Nil(t, err)
+
+	// consume stored messages
+	topics := []string{string(BeginBlockReqTopic), BeginBlockResTopic, StateChangeTopic}
+	msgs, err := poll(bootstrapServers, topics, 5)
+	require.Nil(t, err)
+
+	// validate data stored in Kafka
+	require.Equal(t, expectedBeginBlockReqBytes, getMessageValueForTopic(msgs, string(BeginBlockReqTopic), 0))
```

Comment:
I can't get these to pass.

```console
$ docker-compose -f plugin/plugins/kafka/docker-compose.yml up -d zookeeper broker
[+] Running 3/3
 ⠿ Network kafka_default  Created                                                                                                                0.0s
 ⠿ Container zookeeper    Started                                                                                                                0.4s
 ⠿ Container broker       Started                                                                                                                1.0s
$ go test -mod=readonly -tags "dynamic cgo ledger test_ledger_mock norace" ./... --run 'Kafka'
<snip>
--- FAIL: TestKafkaStreamingService (53.45s)
    service_test.go:217:
        	Error Trace:	service_test.go:217
        	            				service_test.go:165
        	Error:      	Not equal:
        	            	expected: []byte{0x7b, 0x22, 0x68, 0x61, 0x73, 0x68, 0x22, 0x3a, 0x22, 0x41, 0x51, 0x49, 0x44, 0x42, 0x41, 0x55, 0x47, 0x42, 0x77, 0x67, 0x4a, 0x22, 0x2c, 0x22, 0x68, 0x65, 0x61, 0x64, 0x65, 0x72, 0x22, 0x3a, 0x7b, 0x22, 0x76, 0x65, 0x72, 0x73, 0x69, 0x6f, 0x6e, 0x22, 0x3a, 0x7b, 0x22, 0x62, 0x6c, 0x6f, 0x63, 0x6b, 0x22, 0x3a, 0x22, 0x30, 0x22, 0x2c, 0x22, 0x61, 0x70, 0x70, 0x22, 0x3a, 0x22, 0x30, 0x22, 0x7d, 0x2c, 0x22, 0x63, 0x68, 0x61, 0x69, 0x6e, 0x5f, 0x69, 0x64, 0x22, 0x3a, 0x22, 0x22, 0x2c, 0x22, 0x68, 0x65, 0x69, 0x67, 0x68, 0x74, 0x22, 0x3a, 0x22, 0x31, 0x22, 0x2c, 0x22, 0x74, 0x69, 0x6d, 0x65, 0x22, 0x3a, 0x22, 0x30, 0x30, 0x30, 0x31, 0x2d, 0x30, 0x31, 0x2d, 0x30, 0x31, 0x54, 0x30, 0x30, 0x3a, 0x30, 0x30, 0x3a, 0x30, 0x30, 0x5a, 0x22, 0x2c, 0x22, 0x6c, 0x61, 0x73, 0x74, 0x5f, 0x62, 0x6c, 0x6f, 0x63, 0x6b, 0x5f, 0x69, 0x64, 0x22, 0x3a, 0x7b, 0x22, 0x68, 0x61, 0x73, 0x68, 0x22, 0x3a, 0x6e, 0x75, 0x6c, 0x6c, 0x2c, 0x22, 0x70, 0x61, 0x72, 0x74, 0x5f, 0x73, 0x65, 0x74, 0x5f, 0x68, 0x65, 0x61, 0x64, 0x65, 0x72, 0x22, 0x3a, 0x7b, 0x22, 0x74, 0x6f, 0x74, 0x61, 0x6c, 0x22, 0x3a, 0x30, 0x2c, 0x22, 0x68, 0x61, 0x73, 0x68, 0x22, 0x3a, 0x6e, 0x75, 0x6c, 0x6c, 0x7d, 0x7d, 0x2c, 0x22, 0x6c, 0x61, 0x73, 0x74, 0x5f, 0x63, 0x6f, 0x6d, 0x6d, 0x69, 0x74, 0x5f, 0x68, 0x61, 0x73, 0x68, 0x22, 0x3a, 0x6e, 0x75, 0x6c, 0x6c, 0x2c, 0x22, 0x64, 0x61, 0x74, 0x61, 0x5f, 0x68, 0x61, 0x73, 0x68, 0x22, 0x3a, 0x6e, 0x75, 0x6c, 0x6c, 0x2c, 0x22, 0x76, 0x61, 0x6c, 0x69, 0x64, 0x61, 0x74, 0x6f, 0x72, 0x73, 0x5f, 0x68, 0x61, 0x73, 0x68, 0x22, 0x3a, 0x6e, 0x75, 0x6c, 0x6c, 0x2c, 0x22, 0x6e, 0x65, 0x78, 0x74, 0x5f, 0x76, 0x61, 0x6c, 0x69, 0x64, 0x61, 0x74, 0x6f, 0x72, 0x73, 0x5f, 0x68, 0x61, 0x73, 0x68, 0x22, 0x3a, 0x6e, 0x75, 0x6c, 0x6c, 0x2c, 0x22, 0x63, 0x6f, 0x6e, 0x73, 0x65, 0x6e, 0x73, 0x75, 0x73, 0x5f, 0x68, 0x61, 0x73, 0x68, 0x22, 0x3a, 0x6e, 0x75, 0x6c, 0x6c, 0x2c, 0x22, 0x61, 0x70, 0x70, 0x5f, 0x68, 0x61, 0x73, 0x68, 0x22, 0x3a, 0x6e, 0x75, 0x6c, 0x6c, 0x2c, 0x22, 0x6c, 0x61, 0x73, 0x74, 0x5f, 0x72, 0x65, 0x73, 0x75, 0x6c, 0x74, 0x73, 0x5f, 0x68, 0x61, 0x73, 0x68, 0x22, 0x3a, 0x6e, 0x75, 0x6c, 0x6c, 0x2c, 0x22, 0x65, 0x76, 0x69, 0x64, 0x65, 0x6e, 0x63, 0x65, 0x5f, 0x68, 0x61, 0x73, 0x68, 0x22, 0x3a, 0x6e, 0x75, 0x6c, 0x6c, 0x2c, 0x22, 0x70, 0x72, 0x6f, 0x70, 0x6f, 0x73, 0x65, 0x72, 0x5f, 0x61, 0x64, 0x64, 0x72, 0x65, 0x73, 0x73, 0x22, 0x3a, 0x6e, 0x75, 0x6c, 0x6c, 0x7d, 0x2c, 0x22, 0x6c, 0x61, 0x73, 0x74, 0x5f, 0x63, 0x6f, 0x6d, 0x6d, 0x69, 0x74, 0x5f, 0x69, 0x6e, 0x66, 0x6f, 0x22, 0x3a, 0x7b, 0x22, 0x72, 0x6f, 0x75, 0x6e, 0x64, 0x22, 0x3a, 0x31, 0x2c, 0x22, 0x76, 0x6f, 0x74, 0x65, 0x73, 0x22, 0x3a, 0x5b, 0x5d, 0x7d, 0x2c, 0x22, 0x62, 0x79, 0x7a, 0x61, 0x6e, 0x74, 0x69, 0x6e, 0x65, 0x5f, 0x76, 0x61, 0x6c, 0x69, 0x64, 0x61, 0x74, 0x6f, 0x72, 0x73, 0x22, 0x3a, 0x5b, 0x5d, 0x7d}
        	            	actual  : []byte{0xa, 0x9, 0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7, 0x8, 0x9, 0x12, 0x15, 0xa, 0x0, 0x18, 0x1, 0x22, 0xb, 0x8, 0x80, 0x92, 0xb8, 0xc3, 0x98, 0xfe, 0xff, 0xff, 0xff, 0x1, 0x2a, 0x2, 0x12, 0x0, 0x1a, 0x2, 0x8, 0x1}

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,32 +1,5 @@
        	            	-([]uint8) (len=465) {
        	            	- 00000000  7b 22 68 61 73 68 22 3a  22 41 51 49 44 42 41 55  |{"hash":"AQIDBAU|
        	            	- 00000010  47 42 77 67 4a 22 2c 22  68 65 61 64 65 72 22 3a  |GBwgJ","header":|
        	            	- 00000020  7b 22 76 65 72 73 69 6f  6e 22 3a 7b 22 62 6c 6f  |{"version":{"blo|
        	            	- 00000030  63 6b 22 3a 22 30 22 2c  22 61 70 70 22 3a 22 30  |ck":"0","app":"0|
        	            	- 00000040  22 7d 2c 22 63 68 61 69  6e 5f 69 64 22 3a 22 22  |"},"chain_id":""|
        	            	- 00000050  2c 22 68 65 69 67 68 74  22 3a 22 31 22 2c 22 74  |,"height":"1","t|
        	            	- 00000060  69 6d 65 22 3a 22 30 30  30 31 2d 30 31 2d 30 31  |ime":"0001-01-01|
        	            	- 00000070  54 30 30 3a 30 30 3a 30  30 5a 22 2c 22 6c 61 73  |T00:00:00Z","las|
        	            	- 00000080  74 5f 62 6c 6f 63 6b 5f  69 64 22 3a 7b 22 68 61  |t_block_id":{"ha|
        	            	- 00000090  73 68 22 3a 6e 75 6c 6c  2c 22 70 61 72 74 5f 73  |sh":null,"part_s|
        	            	- 000000a0  65 74 5f 68 65 61 64 65  72 22 3a 7b 22 74 6f 74  |et_header":{"tot|
        	            	- 000000b0  61 6c 22 3a 30 2c 22 68  61 73 68 22 3a 6e 75 6c  |al":0,"hash":nul|
        	            	- 000000c0  6c 7d 7d 2c 22 6c 61 73  74 5f 63 6f 6d 6d 69 74  |l}},"last_commit|
        	            	- 000000d0  5f 68 61 73 68 22 3a 6e  75 6c 6c 2c 22 64 61 74  |_hash":null,"dat|
        	            	- 000000e0  61 5f 68 61 73 68 22 3a  6e 75 6c 6c 2c 22 76 61  |a_hash":null,"va|
        	            	- 000000f0  6c 69 64 61 74 6f 72 73  5f 68 61 73 68 22 3a 6e  |lidators_hash":n|
        	            	- 00000100  75 6c 6c 2c 22 6e 65 78  74 5f 76 61 6c 69 64 61  |ull,"next_valida|
        	            	- 00000110  74 6f 72 73 5f 68 61 73  68 22 3a 6e 75 6c 6c 2c  |tors_hash":null,|
        	            	- 00000120  22 63 6f 6e 73 65 6e 73  75 73 5f 68 61 73 68 22  |"consensus_hash"|
        	            	- 00000130  3a 6e 75 6c 6c 2c 22 61  70 70 5f 68 61 73 68 22  |:null,"app_hash"|
        	            	- 00000140  3a 6e 75 6c 6c 2c 22 6c  61 73 74 5f 72 65 73 75  |:null,"last_resu|
        	            	- 00000150  6c 74 73 5f 68 61 73 68  22 3a 6e 75 6c 6c 2c 22  |lts_hash":null,"|
        	            	- 00000160  65 76 69 64 65 6e 63 65  5f 68 61 73 68 22 3a 6e  |evidence_hash":n|
        	            	- 00000170  75 6c 6c 2c 22 70 72 6f  70 6f 73 65 72 5f 61 64  |ull,"proposer_ad|
        	            	- 00000180  64 72 65 73 73 22 3a 6e  75 6c 6c 7d 2c 22 6c 61  |dress":null},"la|
        	            	- 00000190  73 74 5f 63 6f 6d 6d 69  74 5f 69 6e 66 6f 22 3a  |st_commit_info":|
        	            	- 000001a0  7b 22 72 6f 75 6e 64 22  3a 31 2c 22 76 6f 74 65  |{"round":1,"vote|
        	            	- 000001b0  73 22 3a 5b 5d 7d 2c 22  62 79 7a 61 6e 74 69 6e  |s":[]},"byzantin|
        	            	- 000001c0  65 5f 76 61 6c 69 64 61  74 6f 72 73 22 3a 5b 5d  |e_validators":[]|
        	            	- 000001d0  7d                                                |}|
        	            	+([]uint8) (len=38) {
        	            	+ 00000000  0a 09 01 02 03 04 05 06  07 08 09 12 15 0a 00 18  |................|
        	            	+ 00000010  01 22 0b 08 80 92 b8 c3  98 fe ff ff ff 01 2a 02  |."............*.|
        	            	+ 00000020  12 00 1a 02 08 01                                 |......|
        	            	 }
        	Test:       	TestKafkaStreamingService
FAIL
FAIL	github.com/cosmos/cosmos-sdk/plugin/plugins/kafka/service	53.923s
<snip>
```

Reply:
I tried changing the `require.Equal` calls to `Asseret.Equal` so that I could see the results of those test points as well. I did that in `testListenBeginBlock`, `testListenDeliverTx1`, `testListenDeliverTx2`, and `testListenEndBlock`. All of those pieces fail for me with similar unreadable messages coming back from `getMessageValueForTopic`.

Relatedly, it's okay to use `require` when the test should immediately stop due to a failure (e.g. when setting up a test), but after setup, when doing several checks (like these msg value comparisons), it's better to use `assert`. That way, you know more about what's breaking and aren't constantly rerunning tests just to find the next thing you need to fix.